### PR TITLE
Extend ogre 1.x custom shaders support

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

Related https://github.com/gazebosim/gz-sim/issues/2002#issuecomment-1718745400

## Summary

Extended to support features needed to render waves:
* 2d textures
* cubemap textures
* constant variables

Code is ported from [the ogre2 implementation](https://github.com/gazebosim/gz-rendering/blob/59417bad192e096b5b3b0a28e28388c66de1252a/ogre2/src/Ogre2Material.cc#L848)

## Test it
Build and run the waves example

```
# ogre 1.x is used when no args are supplied
./waves
```

Sky is not supported yet in ogre 1.x so you'll only see waves in this demo:

ogre_waves

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: examples/waves/media/GerstnerWaves_fs.glsl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+////////// Input computed in vertex shader //////////
+varying mat3 rotMatrix;
+varying vec3 eyeVec;
+varying vec2 bumpCoord;
+
+void main()
+{
+  // Apply bump mapping to normal vector to make waves look more detailed:
+  vec4 bump = texture2D(bumpMap, bumpCoord)*2.0 - 1.0;
+  vec3 N = normalize(rotMatrix * bump.xyz);
```
Comment (ahcorde) [inline examples/waves/media/GerstnerWaves_fs.glsl:38]:
```suggestion
  vec4 bump = texture2D(bumpMap, bumpCoord) * 2.0 - 1.0;
  vec3 N = normalize(rotMatrix * bump.xyz);
```

File: examples/waves/media/GerstnerWaves_fs.glsl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  // uncomment this line if using other textures that are Y up
+  // Gazebo requires rotated cube map lookup.
+  // R = vec3(R.x, R.z, R.y);
+
+  // Get environment color of reflected ray:
+  vec4 envColor = textureCube(cubeMap, R, 0.0);
+
+  // Cheap hdr effect:
+  envColor.rgb *= (envColor.r+envColor.g+envColor.b)*hdrMultiplier;
```
Comment (ahcorde) [inline examples/waves/media/GerstnerWaves_fs.glsl:55]:
```suggestion
  envColor.rgb *= (envColor.r + envColor.g + envColor.b) * hdrMultiplier;
```

File: ogre/src/OgreMaterial.cc [hunk 3]
```diff
@@ -441,6 +460,84 @@ void OgreMaterial::UpdateShaderParams(ConstShaderParamsPtr _params,
...
+        // https://github.com/gazebosim/gz-rendering/pull/541
+        if (texUnit->isLoaded())
+        {
+          texUnit->_load();
+        }
+      }
+      else
+      {
+        gzerr << "Unrecognized texture type set for shader param: "
+               << name_param.first << std::endl;
```
Comment (ahcorde) [inline ogre/src/OgreMaterial.cc:531]:
```suggestion
        gzerr << "Unrecognized texture type set for shader param: "
              << name_param.first << std::endl;
```

**Review by ahcorde — APPROVED** [review]
Minor style issues. is it possible to add a test ?

File: ogre/src/OgreMaterial.cc (hunk 3 above)
Comment (iche033) [reply ogre/src/OgreMaterial.cc:531]: [88b4fb7](https://github.com/gazebosim/gz-rendering/pull/908/commits/88b4fb7584cc704390ccf20edd788d1b7e0b6aee)

File: examples/waves/media/GerstnerWaves_fs.glsl (hunk 2 above)
Comment (iche033) [reply examples/waves/media/GerstnerWaves_fs.glsl]: [88b4fb7](https://github.com/gazebosim/gz-rendering/pull/908/commits/88b4fb7584cc704390ccf20edd788d1b7e0b6aee)

File: examples/waves/media/GerstnerWaves_fs.glsl (hunk 1 above)
Comment (iche033) [reply examples/waves/media/GerstnerWaves_fs.glsl]: [88b4fb7](https://github.com/gazebosim/gz-rendering/pull/908/commits/88b4fb7584cc704390ccf20edd788d1b7e0b6aee)

**Comment by iche033** [comment]
> is it possible to add a test ?

yep, will add one

**Comment by iche033** [comment]
> yep, will add one

Added `INTEGRATION_waves` test that verifies waves can be rendered. Shaders and materials files for the test are taken from the waves demo. [20ca0c7](https://github.com/gazebosim/gz-rendering/pull/908/commits/20ca0c7c4be6ffa4b2d42d9843fcbce61e3623d6)

File: test/integration/waves.cc [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <gtest/gtest.h>
+
```
Comment (ahcorde) [inline test/integration/waves.cc:19]:
```suggestion
`include <string>`
`include <memory>`
```

**Review by ahcorde — APPROVED** [review]
Just a minor comment

File: test/integration/waves.cc (hunk 4 above)
Comment (iche033) [reply test/integration/waves.cc:19]: added in [2068ae5](https://github.com/gazebosim/gz-rendering/pull/908/commits/2068ae5ddb1a8ff8d70bad08de69bb454109f1a7)

**Pull request merged**